packages/Main/src/Layer/VpcLayer.js

```diff
@@ -0,0 +1,341 @@
+                                + Math.abs(src.bounds[4] - src.bounds[1])) / (2 * src.span);
+                            this.spacing.push(spacing);
+
+                            const root = new EntwinePointTileNode(0, 0, 0, 0, this, -1, i);
+                            root.bbox.min.fromArray(src.boundsConforming, 0);
+                            root.bbox.max.fromArray(src.boundsConforming, 3);
+                            this.roots[i] = root;
+
+                            return root.loadOctree().then(res => resolve(res));
+                        }
+                    });
+                this.loadOctrees.push(promise);
+
+                root.load = () => this.loadOctrees[i].then(res => res.load());
+                this.roots.push(root);
+            });
+            this.ready = true;
+
+            return this.loadOctrees;
+        });
+    }
+
+    // adapted from PointCloudLayer, return are differents
+    preUpdate(context, changeSources) {
+        // See https://cesiumjs.org/hosted-apps/massiveworlds/downloads/Ring/WorldScaleTerrainRendering.pptx
+        // slide 17
+        context.camera.preSSE =
+            context.camera.height /
+                (2 * Math.tan(THREE.MathUtils.degToRad(context.camera.camera3D.fov) * 0.5));
+
+        if (this.material) {
+            this.material.visible = this.visible;
+            this.material.opacity = this.opacity;
+            this.material.depthWrite = false;
+            this.material.size = this.pointSize;
+            this.material.scale = context.camera.preSSE;
+            if (this.material.updateUniforms) {
+                this.material.updateUniforms();
+            }
+        }
+
+        // lookup lowest common ancestor of changeSources
+        let commonAncestor;
+        for (const source of changeSources.values()) {
+            if (source.isCamera || source == this) {
+                // if the change is caused by a camera move, no need to bother
+                // to find common ancestor: we need to update the whole tree:
+                // some invisible tiles may now be visible
+                return this.roots;
+            }
+            if (source.obj === undefined) {
+                continue;
+            }
+            // filter sources that belong to our layer
+            if (source.obj.isPoints && source.obj.layer == this) {
+                if (!commonAncestor) {
+                    commonAncestor = source;
+                } else {
+                    commonAncestor = source.findCommonAncestor(commonAncestor);
+
+                    if (!commonAncestor) {
+                        return this.roots;
+                    }
+                }
+            }
+        }
+
+        if (commonAncestor) {
+            return [commonAncestor];
+        }
+
+        // Start updating from hierarchy root
+        return this.roots;
+    }
+
+    // PointCloudLayer.update separate in 2 parts: update and subUpdate
+    subUpdate(elt, context, layer, bbox) {
```

Comment:
Same for update and subUpdate: there is a lot of shared code with PointCloudLayer, can you factorize them please ?